extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -246,26 +261,148 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
       .then(() =>
         getIdentityToken(this.win, this.getAmpDoc(), super.getConsentPolicy())
       );
-    this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
 
-    if (this.isResponsive_()) {
-      // Attempt to resize to the correct height. The width should already be
-      // 100vw, but is fixed here so that future resizes of the viewport don't
-      // affect it.
-      const viewportSize = this.getViewport().getSize();
-      return this.attemptChangeSize(
-        AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
-          this.autoFormat_,
-          viewportSize,
-          this.element
-        ),
-        viewportSize.width
-      ).catch(() => {});
+    return this.maybeUpgradeToResponsive().then(() => {
+      this.autoFormat_ = this.element.getAttribute('data-auto-format') || '';
+
+      if (this.isResponsive_()) {
+        // Attempt to resize to the correct height. The width should already be
+        // 100vw, but is fixed here so that future resizes of the viewport don't
+        // affect it.
+        const viewportSize = this.getViewport().getSize();
+        return this.attemptChangeSize(
+          AmpAdNetworkAdsenseImpl.getResponsiveHeightForContext_(
+            this.autoFormat_,
+            viewportSize,
+            this.element
+          ),
+          viewportSize.width
+        ).catch(() => {});
+      }
+
+      // This should happen last, as some diversion criteria rely on some of the
+      // preceding logic (specifically responsive logic).
+      this.divertExperiments();
+      this.maybeAddSinglePassExperiment();
+    });
+  }
+
+  /**
+   * Upgrades the ad unit to responsive if there is an opt-in setting in localstorage.
+   * See https://github.com/ampproject/amphtml/issues/23568 for design.
+   * @return {Promise} a promise that resolves when any upgrade is complete.
+   */
+  maybeUpgradeToResponsive() {
+    if (!this.isInAdSizeOptimizationExperimentBranch()) {
+      return Promise.resolve();
+    }
+    // If the ad unit is already responsive, there's nothing to do.
+    // Note that we can't use this.autoFormat_ here because it's not
+    // populated yet.
+    if (this.element.hasAttribute('data-auto-format')) {
+      return Promise.resolve();
     }
-    // This should happen last, as some diversion criteria rely on some of the
-    // preceding logic (specifically responsive logic).
-    this.divertExperiments();
-    this.maybeAddSinglePassExperiment();
+    return (
```

Comment:
Why the paren?

Reply:
It gets automatically added by the linter:
gulp lint --local_changes --fix
Not sure if there's a way around that.